private/buf/buflsp/file.go

```diff
@@ -161,6 +198,107 @@ func (f *file) Update(ctx context.Context, version int32, text string) {
 	f.hasText = true
 }
 
+// FetchSettings refreshes configuration settings for this file.
+//
+// This only needs to happen when the file is open or when the client signals
+// that configuration settings have changed.
+func (f *file) RefreshSettings(ctx context.Context) {
+	settings, err := f.lsp.client.Configuration(ctx, &protocol.ConfigurationParams{
+		Items: []protocol.ConfigurationItem{
+			{ScopeURI: f.uri, Section: "buf.against"},
+			{ScopeURI: f.uri, Section: "buf.gitRemote"},
+		},
+	})
+	if err != nil {
+		// We can throw the error away, since the handler logs it for us.
+		return
+	}
+
+	f.againstKind = getSetting(f, settings, "buf.against", 0, againstKindFromString)
+	f.gitRemote = getSetting(f, settings, "buf.gitRemote", 1, func(s string) (string, bool) { return s, true })
+
+	// Given the remote, we need to find the main branch. First, look up the
+	// remote.
+	remote, err := git.GetRemote(
+		ctx,
+		f.lsp.runner,
+		f.lsp.container,
+		normalpath.Dir(f.uri.Filename()),
+		f.gitRemote,
+	)
+	if errors.Is(err, git.ErrRemoteNotFound) {
+		// This is an invalid remote, so it means gitRemote should be
+		// interpreted as a local branch name instead.
+		f.gitRemoteBranch = f.gitRemote
+		f.gitRemote = ""
+		f.lsp.logger.Debug(
+			"found local branch",
+			slog.String("uri", string(f.uri)),
+			slog.String("ref", f.gitRemoteBranch),
+		)
+	} else if err != nil {
```

Comment:
This behaviour is somewhat reasonable, but I wonder if branch + remote should just be different config fields with sane defaults for clarity. Even if this is documented behaviour, the config field being named `gitRemote` but pointing to a local branch might cause some confusion. This would have cascading changes through the `fileOpener`, etc.

Reply:
I think I'm going to change the semantics as follows:
1. againstKind will be `againstGit` or `againstSaved`.
2. There will be a single config knob for the ref to use for `againstGit`, with a default of `origin/HEAD`.